src/authentication/authentication-hapi-plugin.ts

```diff
@@ -106,9 +94,37 @@ class AuthenticationHapiPlugin extends HapiPlugin {
 
   public async getTeamTokenHandler(request: Hapi.Request, reply: Hapi.IReply) {
     try {
-      const teamId = parseInt(request.paramsArray[0], 10);
-      const teamToken: TeamToken[] = await teamTokenQuery(this.db, undefined, teamId);
-      if (!teamToken.length) {
+      const credentials = request.auth.credentials as AccessToken;
+      const userName = sanitizeUsername(credentials.sub);
+      const isAdmin = await this.isAdmin(userName);
+      const userTeams = await this.gitlab.getUserGroups(userName);
+      const parameterKey = 'teamIdOrName';
+      const teamIdOrName = request.params[parameterKey];
+      let ownTeam: Group | undefined;
+      if (userTeams && teamIdOrName) {
+        ownTeam = userTeams.find(assertTeam.bind(undefined, teamIdOrName));
+      }
+
+      let teamId: number | undefined;
+
+      if (ownTeam) { // token for a team that the user belongs to
+        teamId = ownTeam.id;
+      } else if (isAdmin && teamIdOrName) { // An admin can get any team's token
+        teamId = await this.teamIdOrNameToTeamId(teamIdOrName);
+      } else if (!teamIdOrName) {
+        if (!userTeams.length) {
+          throw Error(`User ${userName} is not in any team`);
+        }
+        if (userTeams.length > 1) {
+          throw Error(`User ${userName} is in multiple teams`);
+        }
+        // NOTE: we only support a single team for now
+        teamId = userTeams[0].id;
+      } else {
+        return reply(Boom.create(401, `Insufficient privileges`));
+      }
```

Comment:
The above flow is a bit hard to follow, but I want to hear the teamIdOrName change reasoning first.